runtime/commonMain/src/kotlinx/serialization/internal/Enums.kt

```diff
@@ -20,45 +24,53 @@ public class EnumDescriptor(override val name: String, private val choices: Arra
 
     override fun equals(other: Any?): Boolean {
         if (this === other) return true
-        if (other == null || this::class != other::class) return false
+        if (other == null) return false
+        if (other !is SerialDescriptor) return false
 
-        other as EnumDescriptor
+        if (other.kind !== UnionKind.ENUM_KIND) return false
 
         if (name != other.name) return false
-        if (!choices.contentEquals(other.choices)) return false
+        if (elementNames() != other.elementNames()) return false
 
         return true
     }
 
+    override fun toString(): String {
+        return elementNames().joinToString(", ", "$name(", ")")
+    }
+
     override fun hashCode(): Int {
         var result = super.hashCode()
         result = 31 * result + name.hashCode()
-        result = 31 * result + choices.contentHashCode()
+        result = 31 * result + values.contentHashCode()
         return result
     }
 }
 
-open class CommonEnumSerializer<T>(val serialName: String, val choices: Array<T>, choicesNames: Array<String>) :
+open class CommonEnumSerializer<T>(serialName: String, val values: Array<T>, valuesNames: Array<String>) :
```

Comment:
Deprecation with error on both classes?

Reply:
Basically, `CommonEnumSerializer` is still used when we don't need a full-blown generated serializer (when no `@SerialInfo` annotations are present). `EnumSerializer` can be deprecated, but it is more convenient to use it when we are on JVM (and therefore have all the information about enum we need). I'll suggest going with `@InternalSerializationApi` on both.